py-monty/pymatgen/spglib: bump versions and fix py-pymatgen dependencies

bernhardkaindl opened the pull request:
Fix build of old versions to use pytong 3.6 or older: They are pre-cythonized so can't build with python@3.7:
Add version 2021.3.5  is the last version using the old/revious API,

adamjstewart reviewed (changes requested):
Dependencies need to be updated:
```python
    python_requires=">=3.7",                                                             
    install_requires=[                                                                   
        "numpy>=1.20.1",                                                                 
        "requests",                                                                      
        "ruamel.yaml>=0.15.6",                                                           
        "monty>=3.0.2",                                                                  
        "scipy>=1.5.0",                                                                  
        "tabulate",                                                                      
        "spglib>=1.9.9.44",                                                              
        "networkx>=2.2",                                                                 
        "matplotlib>=1.5",                                                               
        "palettable>=3.1.1",                                                             
        "sympy",                                                                         
        "pandas",                                                                        
        "plotly>=4.5.0",                                                                 
        "uncertainties>=3.1.4",                                                          
    ],                
```

bernhardkaindl commented:
@spackbot run pipeline

bernhardkaindl commented:
> Dependencies need to be updated:

Done. Added the new dependencies and added the version. to get to the versions listed, it was also nedded to:
* Add a current sha to pty-mony and py-spglib.

bernhardkaindl committed:
py-pymatgen: fix build of old versions, bump version to 2021.3.9

Fix the build of old versions: pymatgen@4.7.2 does not compile with python>=3.7

Bump version to 2021.3.9, with required dependencies are bumped as well:
py-monty, py-pymatgen, py-requests, py-ruamel-yaml, py-spglib, py-uncertainties

adamjstewart commented on var/spack/repos/builtin/packages/py-monty/package.py:
```diff
@@ -12,6 +12,7 @@ class PyMonty(PythonPackage):
+    version('2021.8.17', sha256='d4d5b85566bda80360e275e6ffb72228d203de68c5155446a0e09f19c63e8540')
     version('0.9.6', sha256='bbf05646c4e86731c2398a57b1044add7487fc4ad03122578599ddd9a8892780')
 
     depends_on('py-setuptools', type='build')
```
Python 3.5+ is now required, and six is no longer required

bernhardkaindl replied:
added

adamjstewart replied:
Did you forget to push? I don't see these changes

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,27 @@ class PyPymatgen(PythonPackage):
     version('4.7.2', sha256='e439b78cc3833a03963c3c3efe349d8a0e52a1550c8a05c56a89aa1b86657436')
     version('4.6.2', sha256='f34349090c6f604f7d402cb09cd486830b38523639d7160d7fd282d504036a0e')
 
     extends('python', ignore='bin/tabulate')
```
The latest version requires Python 3.7+

bernhardkaindl replied:
added

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,27 @@ class PyPymatgen(PythonPackage):
-    depends_on('py-six',                 type=('build', 'run'))
+    depends_on('py-cython',              type=('build'))
+    depends_on('py-setuptools@18.0:',    type=('build'))
+    depends_on('py-setuptools@18.0:44',  type=('build'),        when='@:4.7.2')
```
Where does this constraint come from?

bernhardkaindl replied:
Sorry, my bad, removed.

bernhardkaindl commented:
@tldahlgren I applied the dependency updates as required, two dependencies need new versions for this as well.

Can you approve? Thanks!

tldahlgren reviewed (dismissed):
Confirmed the new versions' sha256.

tldahlgren commented:
@adamjstewart I believe your re-review is required before this PR can be merged.

bernhardkaindl commented:
> @adamjstewart I believe your re-review is required before this PR can be merged.

Thanks, yes indeed. I'll add a note-to-self that this will have to wait until he is back.

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,29 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-pandas',              type=('build', 'run'))
+    depends_on('py-plotly@4.5.0:',       type=('build', 'run'))
+    depends_on('py-uncertainties@3.1.4', type=('build', 'run'))
+    depends_on('py-netcdf4',             type=('test', 'run'))
```
This dependency appears to be optional. I would either remove it or add a variant that defaults to False

adamjstewart reviewed (changes requested):
Sorry for the wait! This PR just needs updates to `py-monty` and a unneeded dep in `py-pymatgen` and then it can be merged.

bernhardkaindl commented:
@adamjstewart py-monty (and others) are updated to the required versions (as per requirements.txt)
And as requested, I made `netcdf4` and `networkx` optional variants as they are optional dependencies.

adamjstewart commented:
The `py-monty` package still hasn't been updated

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
 
-    extends('python', ignore='bin/tabulate')
+    # starting from version 2021.1.1, pymatgen requires Python >=3.7
+    extends('python@:3.6',               ignore='bin/tabulate',  when='@:4.7.2')
```
```suggestion
    extends('python',               ignore='bin/tabulate',  when='@:4.7.2')
```
There are no restrictions on the Python version for older releases

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
-    depends_on('py-six',                 type=('build', 'run'))
+    depends_on('py-six',                 type=('build', 'run'),  when='@:4.7.2')
     depends_on('py-requests',            type=('build', 'run'))
     depends_on('py-pyyaml@3.11:',        type=('build', 'run'))
```
pyyaml isn't needed in newer releases

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    # dependencies of never versions
+    depends_on('py-matplotlib@3.4.2:',    when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-monty@2021.8.17:',     when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-numpy@1.21.2:',        when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-numpy@1.20.1:',        when='@2021.1.1:', type=('build', 'run'))
```
At least in the latest release you add

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-pandas@1.3.2:',        when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-plotly@4.14.3:',       when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-requests@2.26.0',      when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-ruamel-yaml@0.17.4:',  when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-ruamel-yaml@0.15.6:',  when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+
+    # dependencies of never versions
+    depends_on('py-matplotlib@3.4.2:',    when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-monty@2021.8.17:',     when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-monty@3.0.2:',     when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-plotly@4.14.3:',       when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-requests@2.26.0',      when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-ruamel-yaml@0.17.4:',  when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-scipy@1.7.1:',         when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-scipy@1.5.0:',         when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-requests@2.26.0',      when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-ruamel-yaml@0.17.4:',  when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-scipy@1.7.1:',         when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-spglib@1.16.1:',       when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-spglib@1.9.9.44:',       when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-uncertainties@3.1.5:', when='@2021.1.1:', type=('build', 'run'))
+    # while optional, for imports of spack install --test=root, these are required
+    depends_on('py-netcdf4@1.5.3:',       when='@2021.1.1:', type=('test', 'run'))
+    depends_on('py-networkx@2.5.1:',      when='@2021.1.1:', type=('test', 'run'))
```
```suggestion
    depends_on('py-networkx@2.2:',      when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
     depends_on('py-palettable@2.1.1:',   type=('build', 'run'))
+
+    # dependencies of never versions
+    depends_on('py-matplotlib@3.4.2:',    when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-matplotlib@1.5:',    when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-ruamel-yaml@0.17.4:',  when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-scipy@1.7.1:',         when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-spglib@1.16.1:',       when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-sympy@1.8:',           when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-sympy',           when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-monty@2021.8.17:',     when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-numpy@1.21.2:',        when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-palettable@3.1.1:',    when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-pandas@1.3.2:',        when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-pandas',        when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-numpy@1.21.2:',        when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-palettable@3.1.1:',    when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-pandas@1.3.2:',        when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-plotly@4.14.3:',       when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-plotly@4.5.0:',       when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-scipy@1.7.1:',         when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-spglib@1.16.1:',       when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-sympy@1.8:',           when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-tabulate@0.8.9:',      when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-tabulate',      when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-spglib@1.16.1:',       when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-sympy@1.8:',           when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-tabulate@0.8.9:',      when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-uncertainties@3.1.5:', when='@2021.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('py-uncertainties@3.1.4:', when='@2021.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,43 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-tabulate@0.8.9:',      when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-uncertainties@3.1.5:', when='@2021.1.1:', type=('build', 'run'))
+    # while optional, for imports of spack install --test=root, these are required
+    depends_on('py-netcdf4@1.5.3:',       when='@2021.1.1:', type=('test', 'run'))
```
I would still remove this or create a `abinit` variant and skip those imports, we don't want to add optional dependencies like netcdf or else the build will take longer

adamjstewart commented on var/spack/repos/builtin/packages/py-requests/package.py:
```diff
@@ -12,6 +12,7 @@ class PyRequests(PythonPackage):
+    version('2.26.0', sha256='b8aa58f8cf793ffd8782d3d8cb19e66ef36f7aba4353eec859e74678b01b07a7')
     version('2.25.1', sha256='27973dd4a904a4f13b263a19c866c13b92a39ed1c964655f025f3f8d3d75b804')
     version('2.24.0', sha256='b3559a131db72c33ee969480840fff4bb6dd111de7dd27c8ee1f820f4f00231b')
     version('2.23.0', sha256='b3f43d496c6daba4493e7c431722aeb7dbc6288f52a6e04e7b6023b0247817e6')
```
Dependencies need to be updated to:
```python
requires = [                                                                             
    'charset_normalizer~=2.0.0; python_version >= "3"',                                  
    'chardet>=3.0.2,<5; python_version < "3"',                                           
    'idna>=2.5,<3; python_version < "3"',                                                
    'idna>=2.5,<4; python_version >= "3"',                                               
    'urllib3>=1.21.1,<1.27',                                                             
    'certifi>=2017.4.17'                                                                 
                                                                                         
]  
    python_requires=">=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*, !=3.5.*",       
```

adamjstewart commented on var/spack/repos/builtin/packages/py-ruamel-yaml/package.py:
```diff
@@ -15,6 +15,7 @@ class PyRuamelYaml(PythonPackage):
+    version('0.17.16', sha256='1a771fc92d3823682b7f0893ad56cb5a5c87c48e62b5399d6f42c8759a583b33')
     version('0.16.10', sha256='099c644a778bf72ffa00524f78dd0b6476bca94a1da344130f4bf3381ce5b954')
     version('0.16.5',  sha256='412a6f5cfdc0525dee6a27c08f5415c7fd832a7afcb7a0ed7319628aed23d408')
     version('0.11.7',  sha256='c89363e16c9eafb9354e55d757723efeff8682d05e56b0881450002ffb00a344')
```
Latest version requires Python 3 and setuptools 28.7+

bernhardkaindl commented:
Of course py-pymatgen-4.7.2 does not declare that it doesn't build with python-3.7+, which likely didn't exist when 4.7.2 was tagged.
This happens when trying to build with newer python versions which have a newer API:
```py
427    /home/kaindlb/spack/spack/.opt/linux-ubuntu18.04-zen/gcc-8.4.0/python-3.8.11-5ubyrfelrqpn7gq533v5b4y6h2nywb6x/include/python3.8/cpython/object.h:260:30: note: declared here
428         Py_DEPRECATED(3.8) int (*tp_print)(PyObject *, FILE *, int);
429                                  ^~~~~~~~
430    pymatgen/optimization/linear_assignment.c: In function '__Pyx__ExceptionSave':
431    pymatgen/optimization/linear_assignment.c:21345:21: error: 'PyThreadState' {aka 'struct _ts'} has no member named 'exc_type'; did you mean 'curexc_type'?
```

bernhardkaindl commented:
I've removed the 2022 version from this PR as it would require the newer versions which I added (which I've downgraded on your reuqests again now) to keep this PR *"somewhat"* manageable.

https://libraries.io/pypi/pymatgen/2021.3.9 says that it is content with any version of `requests`, so I'm dropping the update to `py-requests@0.26` from this PR: Better do it as a separate PR, it needs `charset_normalizer` which isn't in spack yet and maybe more.

bernhardkaindl commented:
On the dependency of `py-netcdf4`: As you refer to the build time this causes:

I've rebuilt all python dependencies of `py-pymatgen@4.7.2` (these are much fewer than for the newer versions) with `^python@3.7.11` to check if `py-pymatgen@4.7.2` could be built with python-3.7(answer: It does not, see my comment above) and even that smaller set of dependencies took very long to build (I guess it was like 20-30 minutes or more, I was really wondering if I should have even started that check just to verify for a 2nd time what I already tried) - and this was only the much much smaller list of dependencies of `4.7.2`!. And then I built `py-pymatgen` with py-netcdf4, and despite having to build a few packages, was very quick to build. Also I don't want to cripple the build to a non-standard version just for <5% of build time. I therefore don't see your point on the build time, please reconsider having to invest time into removing the files which import netcdf4, thanks!

adamjstewart commented:
> Of course py-pymatgen-4.7.2 does not declare that it doesn't build with python-3.7+, which likely didn't exist when 4.7.2 was tagged. This happens when trying to build with newer python versions which have a newer API:

It might be good to add a comment specifying that older versions come pre-cythonized and so don't support newer Python versions.

> I've removed the 2022 version from this PR as it would require the newer versions which I added (which I've downgraded on your reuqests again now) to keep this PR "somewhat" manageable.

My above version requests were for the 2022 version, are you sure you weren't getting your version requirements from the `master` branch?

> I therefore don't see your point on the build time, please reconsider having to invest time into removing the files which import netcdf4, thanks!

I'm not suggesting any changes should be made to the library, I'm suggesting changing the files that get imported in the import tests. For example, see the `py-xarray` package where we explicitly define `import_modules` to be all modules _except_ `xarray.tests` which requires additional dependencies to be installed.

bernhardkaindl commented:
> > Of course py-pymatgen-4.7.2 does not declare that it doesn't build with python-3.7+, which likely didn't exist when 4.7.2 was tagged. This happens when trying to build with newer python versions which have a newer API:
> 
> It might be good to add a comment specifying that older versions come pre-cythonized and so don't support newer Python versions.

Yes, good.

> > I've removed the 2022 version from this PR as it would require the newer versions which I added (which I've downgraded on your reuqests again now) to keep this PR "somewhat" manageable.
> 
> My above version requests were for the 2022 version, are you sure you weren't getting your version requirements from the `master` branch?

No, mine were from the requriements.txt of the official tarball of the pymatgen-2022.0.14 release, which extracted to the `pymatgen-2022.0.14/` directory, I just re-checked again. You can also check directly on github from looking at files of the tagged versions: https://github.com/materialsproject/pymatgen/tags

I postponed 2022.x to a later PR anyways as this PR is already far to large and time-consuming, getting all the additional requirements 100% perfectly right demolished the time budget I had hoped for this PR to have by several orders of magnitude already, forget about 2022.x (and I have to business or other interest in it, I just wanted to investigae to fix the build as a "fun" project, updating it I thought, would be a nice bonus for spack (not for me)).

> > I therefore don't see your point on the build time, please reconsider having to invest time into removing the files which import netcdf4, thanks!
> 
> I'm not suggesting any changes should be made to the library, I'm suggesting changing the files that get imported in the import tests. For example, see the `py-xarray` package where we explicitly define `import_modules` to be all modules _except_ `xarray.tests` which requires additional dependencies to be installed.

It looks quite sloppy and a hack in comparison to me. Again, the added build (not run-) time was neglibile for me.

This whitelist would have to be specially crafted for each release range where it is necessary and it is likely that on updates it would be forgotten to re-create it, new features of new versions could potentially not work because of skipping new imports.
If at all, a blacklist would be clearer in stating what it is there for and better for maintainability.

Limiting the check to just some (*especially* a whitelist, not a blacklist!) of imports ***does appear to be inconsistent*** with the otherwise highly exact and uncompromising accuracy in checking dependencies.

bernhardkaindl committed:
Old versions come pre-cythonized and so don't support newer Python

adamjstewart commented:
> No, mine were from the requriements.txt of the official tarball of the pymatgen-2022.0.14 release

`requirements.txt` isn't a good place to look for dependencies since the build system (and pip) don't use it at all. It often includes optional developers tools like pytest or sphinx. See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#important-files for more info.

bernhardkaindl commented:
> > No, mine were from the requriements.txt of the official tarball of the pymatgen-2022.0.14 release
> 
> `requirements.txt` isn't a good place to look for dependencies since the build system (and pip) don't use it at all. It often includes optional developers tools like pytest or sphinx. See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#important-files for more info.

The last sentence of the 2nd paragraph is:
> You should look out for all of these files, as they may all contain important information about package dependencies.

Maybe it should also have a clear statement like you gave now at the end, then there would be no ambiguity:
> `requirements.txt` isn't a good place to look for dependencies since the build system (and pip) don't use it at all. It often includes optional developers tools like pytest or sphinx.

I 've used your versions now and they match the setup.py of the added version, and they match the setup.py. so it's correct for this version: https://github.com/materialsproject/pymatgen/commit/998a4f845cfe2bb2591f1f202a11492881580289
https://github.com/materialsproject/pymatgen/blob/998a4f845cfe2bb2591f1f202a11492881580289/setup.py

adamjstewart commented:
> Maybe it should also have a clear statement like you gave now at the end, then there would be no ambiguity

Later in the docs for finding Python deps we do explicitly say:

> In the root directory of the package, you may notice a requirements.txt file. It may look like this file contains a list of all of the package’s dependencies. Don’t be fooled. This file is used by tools like Travis to install the pre-requisites for the package… and a whole bunch of other things. It often contains dependencies only needed for unit tests, like:
>
> * mock
> * nose
> * pytest
>
> It can also contain dependencies for building the documentation, like sphinx. If you can’t find any information about the package’s dependencies, you can take a look in requirements.txt, but be sure not to add test or documentation dependencies.

But if you want to submit a PR to make these docs more clear that would be much appreciated! Python's build ecosystem keeps changing so rapidly that it's hard to keep these docs up-to-date.

adamjstewart commented on var/spack/repos/builtin/packages/py-monty/package.py:
```diff
@@ -12,7 +12,9 @@ class PyMonty(PythonPackage):
+    version('2021.8.17', sha256='d4d5b85566bda80360e275e6ffb72228d203de68c5155446a0e09f19c63e8540')
     version('0.9.6', sha256='bbf05646c4e86731c2398a57b1044add7487fc4ad03122578599ddd9a8892780')
 
+    depends_on('python@3.5:')
```
```suggestion
    depends_on('python@3.5:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,42 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-scipy@1.5.0:',         when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-spglib@1.9.9.44:',     when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-sympy',                when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-tabulate',             when='@2021.1.1:', type=('build', 'run'))
```
You can actually remove this line, it's redundant now

bernhardkaindl committed:
Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

bernhardkaindl commented:
@adamjstewart - Changes are applied.

adamjstewart commented on var/spack/repos/builtin/packages/py-pymatgen/package.py:
```diff
@@ -15,21 +15,41 @@ class PyPymatgen(PythonPackage):
+    depends_on('py-spglib@1.9.9.44:',     when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-sympy',                when='@2021.1.1:', type=('build', 'run'))
+    depends_on('py-uncertainties@3.1.4:', when='@2021.1.1:', type=('build', 'run'))
+    # while optional, for imports of spack install --test=root, these are required
```
Can you move this comment 1 line down? networkx is actually required, only netcdf4 is optional.

bernhardkaindl committed:
moved comment one line down

adamjstewart reviewed (approved):
Looks good now, thanks for putting up with all of my change requests!
